File: ocaml/leetcode/arrays.ml
(* TOPIC: Array LeetCode Problems | ocaml arrays.ml *)

(* TODO 1: Two Sum — LC #1
   let two_sum nums target =
     (* O(n) using Hashtbl: tbl maps value → index *)
     (* let tbl = Hashtbl.create 16 in List.iteri ... *)  *)

(* TODO 2: Best Time to Buy and Sell Stock — LC #121
   let max_profit prices =
     (* Track min price; max_profit = max(profit, price - min_price) *)
     (* List.fold_left with (min_price, max_profit) as accumulator *)  *)

(* TODO 3: Container With Most Water — LC #11
   let max_area height =
     (* Two pointer; convert to array for O(1) access *)  *)

(* TODO 4: Product of Array Except Self — LC #238
   let product_except_self nums =
     (* Prefix products then suffix products; no division *)  *)

(* TODO 5: Maximum Subarray (Kadane's) — LC #53
   let max_sub_array nums =
     (* List.fold_left with (current_sum, max_sum) accumulator *)  *)

(* TODO 6: Merge Intervals — LC #56
   let merge intervals =
     (* List.sort then merge overlapping *)  *)

(* TODO 7: Find Minimum in Rotated Sorted Array — LC #153
   let find_min nums =
     (* Binary search on array; if nums[mid] > nums[right] min is right *)  *)

let () = print_endline "TODO: implement array problems"

File: cpp/leetcode/arrays.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
// TOPIC: Arrays — LeetCode | g++ -std=c++20 -o out arrays.cpp && ./out

int main() { std::cout << "Arrays LeetCode — TODO: implement\n"; return 0; }

// TODO 1: Two Sum — LeetCode #1
//   vector<int> twoSum(vector<int>& nums, int target)
//   unordered_map<int,int>: store complement → index. O(n).

// TODO 2: Best Time to Buy and Sell Stock — LeetCode #121
//   int maxProfit(vector<int>& prices)
//   Track minPrice; maxProfit = max(maxProfit, price - minPrice). O(n).

// TODO 3: Product of Array Except Self — LeetCode #238
//   vector<int> productExceptSelf(vector<int>& nums)
//   Prefix pass then suffix pass. No division. O(n) O(1) extra.

// TODO 4: Maximum Subarray — LeetCode #53
//   int maxSubArray(vector<int>& nums)
//   Kadane's algorithm. O(n).

// TODO 5: Merge Intervals — LeetCode #56
//   vector<vector<int>> merge(vector<vector<int>>& intervals)
//   Sort by start; merge overlapping. O(n log n).

// TODO 6: Find Minimum in Rotated Sorted Array — LeetCode #153
//   int findMin(vector<int>& nums)
//   Binary search comparing mid vs right. O(log n).

// TODO 7: Container With Most Water — LeetCode #11
//   int maxArea(vector<int>& height)
//   Two pointers at ends; move shorter side inward. O(n).

// TODO 8: Trapping Rain Water — LeetCode #42
//   int trap(vector<int>& height)
//   Two pointers; track maxLeft, maxRight; accumulate water. O(n) O(1).
